pinot-core/src/main/java/com/linkedin/pinot/core/common/DataFetcher.java

```diff
@@ -0,0 +1,205 @@
+
+import com.linkedin.pinot.core.indexsegment.IndexSegment;
+import com.linkedin.pinot.core.segment.index.readers.Dictionary;
+import java.util.HashMap;
+import java.util.Map;
+
+
+/**
+ * DataFetcher is a higher level abstraction for data fetching. Given an index segment, DataFetcher can manage the
+ * DataSource, Dictionary, BlockValSet and BlockValIterator for this segment, preventing redundant construction for
+ * these instances. DataFetcher can be used by both selection, aggregation and group-by data fetching process, reducing
+ * duplicate codes and garbage collection.
+ */
+public class DataFetcher {
+  private final IndexSegment _indexSegment;
+
+  private final Map<String, DataSource> _columnToDataSourceMap = new HashMap<>();
+  private final Map<String, Dictionary> _columnToDictionaryMap = new HashMap<>();
+  private final Map<String, BlockValSet> _columnToBlockValSetMap = new HashMap<>();
+
+  /**
+   * Constructor for DataFetcher.
+   *
+   * @param indexSegment index segment.
+   */
+  public DataFetcher(IndexSegment indexSegment) {
+    _indexSegment = indexSegment;
+  }
+
+  /**
+   * Given a column, fetch its data source.
+   *
+   * @param column column name.
+   * @return data source associated with this column.
+   */
+  public DataSource getDataSourceForColumn(String column) {
+    DataSource dataSource = _columnToDataSourceMap.get(column);
+    if (dataSource == null) {
+      dataSource = _indexSegment.getDataSource(column);
+      _columnToDataSourceMap.put(column, dataSource);
+    }
+    return dataSource;
+  }
+
+  /**
+   * Given a column, fetch its dictionary.
+   *
+   * @param column column name.
+   * @return dictionary associated with this column.
+   */
+  public Dictionary getDictionaryForColumn(String column) {
+    Dictionary dictionary = _columnToDictionaryMap.get(column);
+    if (dictionary == null) {
+      dictionary = getDataSourceForColumn(column).getDictionary();
+      _columnToDictionaryMap.put(column, dictionary);
+    }
+    return dictionary;
+  }
+
+  /**
+   * Given a column, fetch its block value set.
+   *
+   * @param column column name.
+   * @return block value set associated with this column.
+   */
+  public BlockValSet getBlockValSetForColumn(String column) {
+    BlockValSet blockValSet = _columnToBlockValSetMap.get(column);
+    if (blockValSet == null) {
+      blockValSet = getDataSourceForColumn(column).getNextBlock().getBlockValueSet();
+      _columnToBlockValSetMap.put(column, blockValSet);
+    }
+    return blockValSet;
+  }
+
+  /**
+   * Fetch the dictionary Ids for a single value column.
+   *
+   * @param column column name.
+   * @param inDocIds document Id array.
+   * @param inStartPos input start position.
+   * @param length input length.
+   * @param outDictIds dictionary Id array buffer.
+   * @param outStartPos output start position.
+   */
+  public void fetchSingleDictIds(String column, int[] inDocIds, int inStartPos, int length, int[] outDictIds,
```

Comment:
The DataFetcher serves for both single value and multi value columns. But for multi value columns, we should call the getBlockValSetForColumn() method instead.